pkg/cloudprovider/aws/pricing.go

```diff
@@ -0,0 +1,313 @@
+type PricingProvider struct {
+	ec2     ec2iface.EC2API
+	pricing pricingiface.PricingAPI
+	region  string
+
+	mu                 sync.RWMutex
+	onDemandUpdateTime time.Time
+	onDemandPrices     map[string]float64
+	spotUpdateTime     time.Time
+	spotPrices         map[string]float64
+}
+
+// pricingUpdatePeriod is how often we try to update our pricing information after the initial update on startup
+const pricingUpdatePeriod = 12 * time.Hour
+
+// NewPricingAPI returns a pricing API configured based on a particular region
+func NewPricingAPI(sess *session.Session, region string) pricingiface.PricingAPI {
+	if sess == nil {
+		return nil
+	}
+	// pricing API doesn't have an endpoint in all regions
+	pricingAPIRegion := "us-east-1"
+	if strings.HasPrefix(region, "cn-") {
+		pricingAPIRegion = "cn-north-1"
+	}
+	return pricing.New(sess, &aws.Config{Region: aws.String(pricingAPIRegion)})
+}
+
+func NewPricingProvider(ctx context.Context, pricing pricingiface.PricingAPI, ec2Api ec2iface.EC2API, region string) *PricingProvider {
+	p := &PricingProvider{
+		region:             region,
+		onDemandUpdateTime: initialPriceUpdate,
+		onDemandPrices:     initialOnDemandPrices,
+		spotUpdateTime:     initialPriceUpdate,
+		spotPrices:         initialSpotPrices,
+		ec2:                ec2Api,
+		pricing:            pricing,
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("pricing"))
+
+	// perform an initial price update at startup
+	p.updatePricing(ctx)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				logging.FromContext(ctx).Infof("shutting down")
+				return
+			case <-time.After(pricingUpdatePeriod):
+				p.updatePricing(ctx)
+			}
+		}
+	}()
+	return p
+}
+
+// InstanceTypes returns the list of all instance types for which either a spot or on-demand price is known.
+func (p *PricingProvider) InstanceTypes() []string {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return lo.Union(lo.Keys(p.onDemandPrices), lo.Keys(p.spotPrices))
+}
+
+// OnDemandPrice returns the last known on-demand price for a given instance type, returning an error if there is no
+// known on-demand pricing for the instance type.
+func (p *PricingProvider) OnDemandPrice(instanceType string) (float64, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	price, ok := p.onDemandPrices[instanceType]
+	if !ok {
+		return 0.0, fmt.Errorf("instance type %s not found", instanceType)
+	}
+	return price, nil
+}
+
+// SpotPrice returns the last known spot price for a given instance type, returning an error if there is no
+// known spot pricing for the instance type.
+func (p *PricingProvider) SpotPrice(instanceType string) (float64, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	price, ok := p.spotPrices[instanceType]
+	if !ok {
+		return 0.0, fmt.Errorf("instance type %s not found", instanceType)
+	}
+	return price, nil
+}
+
+func (p *PricingProvider) updatePricing(ctx context.Context) {
+	logging.FromContext(ctx).Infof("updating pricing information")
```

Comment:
Most of our logs start with an Upper case letter. Are you advocating for a lower case pattern? Should we consider migrating the rest of the codebase (over time) to this new pattern?

This line needs to be more explicit about EC2 I think. Look at other log lines and try to keep the concepts aligned
```
karpenter-54564ccf6-dqtpv controller 2022-06-20T21:56:57.492Z	DEBUG	controller.provisioning.cloudprovider	Discovered security groups: [sg-000d68e67c3fe53d8]	{"commit": "cd5f75f", "provisioner": "default"}
karpenter-54564ccf6-dqtpv controller 2022-06-20T21:56:57.494Z	DEBUG	controller.provisioning.cloudprovider	Discovered kubernetes version 1.22	{"commit": "cd5f75f", "provisioner": "default"}
karpenter-54564ccf6-dqtpv controller 2022-06-20T21:56:57.600Z	DEBUG	controller.provisioning.cloudprovider	Discovered ami-04db1f008385aae5e for query "/aws/service/eks/optimized-ami/1.22/amazon-linux-2/recommended/image_id"	{"commit": "cd5f75f", "provisioner": "default"}
```

Reply:
I don't consider either to be better than the other. Initial capitalization of a log is the level of detail I don't think it's worth us having an opinion on.